src/reader.cpp

```diff
@@ -1206,6 +1233,58 @@ void reader_data_t::maybe_refilter_pager(const editable_line_t *el) {
     }
 }
 
+static history_pager_result_t history_pager_search(const std::shared_ptr<history_t> &history,
+                                                   history_search_direction_t direction,
+                                                   size_t history_index,
+                                                   const wcstring &search_string) {
+    constexpr size_t history_pager_max_size = 10;
+    completion_list_t completions;
+    history_search_t search{history, search_string, history_search_type_t::contains,
+                            smartcase_flags(search_string), history_index};
+    while (completions.size() < history_pager_max_size && search.go_to_next_match(direction)) {
+        const history_item_t &item = search.current_item();
+        completions.push_back(
+            completion_t{item.str(), L"", string_fuzzy_match_t::exact_match(),
+                         COMPLETE_REPLACES_TOKEN | COMPLETE_DONT_ESCAPE | COMPLETE_DONT_SORT});
+    }
+    if (direction == history_search_direction_t::forward)
+        std::reverse(completions.begin(), completions.end());
+    return {search_string, completions, search.current_index()};
+}
+
+void reader_data_t::fill_history_pager(bool new_search, history_search_direction_t direction) {
+    assert(!new_search || direction == history_search_direction_t::backward);
+    size_t index;
+    if (new_search) {
+        index = 0;
+    } else if (direction == history_search_direction_t::forward) {
+        index = history_pager_history_index_start;
+    } else {
+        assert(direction == history_search_direction_t::backward);
+        index = history_pager_history_index_end;
+    }
+    // Paranoid copy to avoid data races.
+    // NOLINTNEXTLINE(performance-unnecessary-copy-initialization)
+    wcstring search_term{pager.search_field_line.text()};
+    auto shared_this = this->shared_from_this();
+    debounce_history_pager().perform(
+        [=]() { return history_pager_search(history, direction, index, search_term); },
+        [shared_this, direction, index](const history_pager_result_t &result) {
+            if (result.search_term != shared_this->pager.search_field_line.text())
+                return;  // Stale request.
```

Comment:
```suggestion
        [shared_this, direction, index, new_search](const history_pager_result_t &result) {
            if (result.search_term != shared_this->pager.search_field_line.text())
                return;  // Stale request.
                
            if (result.matched_commands.empty() && !new_search) {
                // No more matches, keep the existing ones and flash.
                shared_this->flash();
                return;
            }
```

That should make it flash if the next "page" doesn't exist, but not if it's a new search.